homeassistant/components/media_player/epson.py

```diff
@@ -0,0 +1,316 @@
+            SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK
+
+EPSON_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Optional(ATTR_CMODE): cv.string
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Epson media player platform."""
+    if DATA_EPSON not in hass.data:
+        hass.data[DATA_EPSON] = []
+
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
+    epson = EpsonProjector(
+        hass, name, host,
+        config.get(CONF_PORT), config.get(CONF_SSL), KEY_COMMANDS)
+
+    epson.update()
+    hass.data[DATA_EPSON].append(epson)
+    async_add_devices([epson], update_before_add=True)
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle for services."""
+        entity_ids = service.data.get('entity_id')
+        if entity_ids:
+            devices = [device for device in hass.data[DATA_EPSON]
+                       if device.entity_id in entity_ids]
+        else:
+            devices = hass.data[DATA_EPSON]
+        for device in devices:
+            if service.service == ATTR_CMODE:
+                cmode = service.data.get(ATTR_CMODE).lower()
+                if cmode in CMODE_LIST_SET:
+                    yield from device.select_cmode(cmode)
+            yield from device.update()
+    hass.services.async_register(
+        DOMAIN, ATTR_CMODE, async_service_handler,
+        schema=EPSON_SCHEMA)
+    return True
+
+
+class EpsonProjector(MediaPlayerDevice):
+    """Representation of Epson Projector Device."""
+
+    def __init__(self, hass, name, host, port, encryption, key_commands):
+        """Initialize entity to control Epson projector."""
+        self._hass = hass
+        self._name = name
+        self._host = host
+        self._port = port
+        self._cmode = None
+        self._source_list = list(DEFAULT_SOURCES.values())
+        self.key_commands = key_commands
+        self._encryption = encryption
+        self._state = STATE_UNKNOWN
+        http_protocol = 'https' if self._encryption else 'http'
+        self._http_url = '{http_protocol}://{host}:{port}/cgi-bin/'.format(
+            http_protocol=http_protocol,
+            host=self._host,
+            port=self._port)
+        referer = "{http_protocol}://{host}:{port}/cgi-bin/webconf".format(
+            http_protocol=http_protocol,
+            host=self._host,
+            port=self._port)
+        self._headers = {
+            "Accept-Encoding": ACCEPT_ENCODING,
+            "Accept": ACCEPT_HEADER,
+            "Referer": referer
+        }
+        self.websession = async_create_clientsession(
+            hass,
+            verify_ssl=self._encryption)
+        self.websession_action = async_create_clientsession(
+            hass,
+            verify_ssl=self._encryption)
```

Comment:
Any reason not to share one session?

Reply:
if I share one session and turn on the projector I get error on update status.
Am I doing something wrong with sharing one session?

```
2018-03-27 11:51:31 WARNING (MainThread) [homeassistant.helpers.entity] Update of media_player.epson_projector is taking over 10 seconds
2018-03-27 11:51:31 INFO (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: timestamp=1522144291.5457628, level=WARNING, message=Update of media_player.epson_projector is taking over 10 seconds, exception=, source=core.py>
2018-03-27 11:51:31 ERROR (MainThread) [custom_components.media_player.epson] Error getting info
```

If I have two sessions it looks like this. No error updating info from projector.

```
2018-03-27 11:56:11 WARNING (MainThread) [homeassistant.helpers.entity] Update of media_player.epson_projector is taking over 10 seconds
2018-03-27 11:56:11 INFO (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: timestamp=1522144571.9009275, level=WARNING, message=Update of media_player.epson_projector is taking over 10 seconds, exception=, source=core.py>
2018-03-27 11:56:16 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_executed[L]: service_call_id=139837689363424-3>
2018-03-27 11:56:59 INFO (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=media_player.epson_projector, old_state=<state media_player.epson_projector=off; cmode=Cinema, friendly_name=Epson Projector, supported_features=36345 @ 2018-03-27T11:55:42.101902+02:00>, new_state=<state media_player.epson_projector=on; source_list=['HDMI1', 'HDMI2', 'PC', 'VIDEO', 'USB', 'LAN', 'WiFi Direct'], cmode=Cinema, friendly_name=Epson Projector, supported_features=36345 @ 2018-03-27T11:56:59.057799+02:00>>
```